grammarinator/generate.py

```diff
@@ -126,12 +121,19 @@ def create_new_test(self, index, lock):
             logger.warning('Test generation failed.', exc_info=e)
             return self.create_new_test(index, lock)
 
-        test_fn = self.out_format % index if self.out_format else None
+        test_fn = self.out_format % index if '%d' in self.out_format else self.out_format
         tree.root = Generator.transform(tree.root, self.transformers)
 
         tree_fn = None
         if self.population and self.keep_trees:
-            tree_fn = join(self.population.directory, basename(test_fn) + Tree.extension)
+            if test_fn:
+                tree_basename = basename(test_fn)
+                if '%d' not in tree_basename:
+                    base, ext = splitext(tree_basename)
+                    tree_basename = '{base}%d{ext}'.format(base=base, ext=ext)
```

Comment:
Is an `if test_fn` check needed here? `basename('')` gives `''`, `splitext('')` gives `('', '')`. So, `tree_basename` will be `'%d'` anyway.

Reply:
Indeed, it's equivalent and simpler. @aroddick could you update it according to this, please?